### PR TITLE
Add linked data entities.

### PR DESCRIPTION
**Pull request by jason-fox**
The whole point of **NGSI-LD** is to be able to traverse the nodes of the knowledge graph. At the moment the IoT Agents just update a single **Device** entity. It is possible to provision relationships in static attributes, but the relationship is directional.

This PR adds the option to update both entities of a relationship  when a measure arrives at the IoT Agent.

`static_attributes` may be supplied with an additional `link` data element when provisioning an IoT Agent to ensure that active attributes from the provisioned IoT Device may be maintained in parallel with a linked data entity . Take for example a temperature gauge placed within a building. The **Device** data model literally represents the IoT device itself, but the `temperature` attribute also needs to be shared with the **Building** entity. A `link` between them can be provisioned.

For example, imagine a Cow wearing an animal collar. The Animal Collar is a **Device** (with `temperature`, `batteryLevel` etc.) and the Cow is an **Animal**  (with `age`, `sex`, `weight` `temperature` etc.). The `temperature` attribute is common to both and should be added to the **Animal** as shown:

```json
"temperature": {
    "type": "Property",
    "value": 21,
    "unitCode": "CEL",
    "providedBy": {
        "type": "Relationship",
        "object": "urn:ngsi-ld:Device:animal-collar2451"
    }
}
```
 
Currently this change has only been made on NGSI-LD entities. It could easily be backported to NGSI-v2 entities if required.

-   Unit Test added
-   Documentation Added

### CODE REVIEW

File: lib/services/ngsi/entities-NGSI-LD.js [hunk 1]
```diff
@@ -383,6 +383,56 @@ function sendQueryValueNgsiLD(entityName, attributes, typeInformation, token, ca
...
+ * @param {Object} json             Payload to send to the context broker
+ */
+function addLinkedEntities(typeInformation, json) {
+    // use the supplied type if possible, but fallback to reading the
+    // type from the id of the entity.
+    function getEntityType(id, target) {
+        if (target.type) {
+            return target.type;
+        }
+        let calculatedType = id.replace(NGSI_LD_URN, '');
```
Comment (bobeal) [inline lib/services/ngsi/entities-NGSI-LD.js:401]: as it is based on a recommendation from the specification, maybe it worth adding a log message (something like `warning, fallback to guessing the type from id as no type attribute was found ...`)?

File: lib/services/ngsi/entities-NGSI-LD.js [hunk 2]
```diff
@@ -383,6 +383,56 @@ function sendQueryValueNgsiLD(entityName, attributes, typeInformation, token, ca
...
+    if(typeInformation.staticAttributes){
+        typeInformation.staticAttributes.forEach((item) => {
+            if (item.link) {
+                linkedTypeInfo[item.name] = item.link;
+            }
+        });
+    }
+
+    Object.keys(linkedTypeInfo).forEach((sourceAttr) => {
+        const sourceAttribute = json[0][sourceAttr];
```
Comment (bobeal) [inline lib/services/ngsi/entities-NGSI-LD.js:416]: are we guaranteed to find the Device payload in the 1st element of the array?

File: test/unit/ngsi-ld/examples/contextRequests/updateContextStaticLinkedAttributes.json [hunk 3]
```diff
@@ -0,0 +1,40 @@
+[
+  {
+    "@context": "http://context.json-ld",
+    "luminosity": {
+      "type": "Property",
+      "value": 87,
+      "providedBy": {
```
Comment (bobeal) [inline test/unit/ngsi-ld/examples/contextRequests/updateContextStaticLinkedAttributes.json:7]: are you sure of this "self relationship"?

File: lib/services/ngsi/entities-NGSI-LD.js (hunk 2 above)
Comment (jason-fox) [reply lib/services/ngsi/entities-NGSI-LD.js:416]: For NGSI-LD, this is handled by `sendUpdateValueNgsiLD` - the existing call is an `upsert`:

```javascript
const url = '/ngsi-ld/v1/entityOperations/upsert/';

    const options = NGSIUtils.createRequestObject(url, typeInformation, token);
    options.method = 'POST';
```

The NGSI-LD Upsert is a batch operation and only accepts an array of elements, hence the 1st element must be a measure from a device.

File: test/unit/ngsi-ld/examples/contextRequests/updateContextStaticLinkedAttributes.json (hunk 3 above)
Comment (jason-fox) [reply test/unit/ngsi-ld/examples/contextRequests/updateContextStaticLinkedAttributes.json:7]: The test defines:

```javascript
  deepClonedObject.types.Lamp.staticAttributes[0].link = {
                attributes: ['luminosity'],
                name: 'providedBy',
                type: 'Building'
            };
```

`staticAttributes[0]` is the `controlledAsset`, so it now has a `link` subproperty.

Therefore when the `luminosity` is set to `87`:

```
{
    "@context": "http://context.json-ld",
    "id": "urn:ngsi-ld:Building:001",
    "type": "Building",
    "luminosity": {
      "type": "Property",
      "value": 87,
      "providedBy": {
        "type": "Relationship",
        "object": "urn:ngsi-ld:Lamp:lamp1"
      }
    }
```

Is upserting the linked entity - i.e. the `urn:ngsi-ld:Building:001` altering the  `luminosity`  attribute and maintaining a `providedBy` relationship of a property.

File: lib/services/ngsi/entities-NGSI-LD.js (hunk 1 above)
Comment (jason-fox) [reply lib/services/ngsi/entities-NGSI-LD.js:401]: This substitution is already made silently in two places:

* https://github.com/telefonicaid/iotagent-node-lib/blob/2835d6d50a27a2406ca316568c8ea60b4f7cddde/lib/services/devices/registrationUtils.js#L467

* https://github.com/telefonicaid/iotagent-node-lib/blob/a7eb2605d342ed2b50821d92f967d93407e90e65/lib/services/ngsi/entities-NGSI-LD.js#L198

The question to @fgalan and the other owners of the repo is whether this should the codebase:

*  have logging in all three places, for consistency 
*  have no logging for any of these instances 

Once this is clear I'll follow up either way.

Comment (fgalan) [reply lib/services/ngsi/entities-NGSI-LD.js:401]: I gusss that a debug trace could be useful and doesn't hurt... so the "have debug logging in all three places, for consistency" option should be the better one.

Comment (jason-fox) [reply lib/services/ngsi/entities-NGSI-LD.js:405]: Fixed 64c7eae

**Comment by mapedraza** [comment]
@bobeal, could you resolve the topics you opened? this PR is waiting your feedback

File: lib/services/ngsi/entities-NGSI-LD.js (hunk 1 above)
Comment (bobeal) [reply lib/services/ngsi/entities-NGSI-LD.js:405]: LGTM

File: lib/services/ngsi/entities-NGSI-LD.js (hunk 2 above)
Comment (bobeal) [reply lib/services/ngsi/entities-NGSI-LD.js:421]: LGTM

File: test/unit/ngsi-ld/examples/contextRequests/updateContextStaticLinkedAttributes.json (hunk 3 above)
Comment (bobeal) [reply test/unit/ngsi-ld/examples/contextRequests/updateContextStaticLinkedAttributes.json:7]: LGTM

**Comment by bobeal** [comment]
> @bobeal, could you resolve the topics you opened? this PR is waiting your feedback

Sorry for the delay!

It seems I'm missing some privileges to resolve the conversations I've opened (seems weird). So I have LGTMed them.

**Comment by fgalan** [comment]
> It seems I'm missing some privileges to resolve the conversations I've opened (seems weird). So I have LGTMed them.

@bobeal so from your perspective the PR is ready to be merged?

**Comment by bobeal** [comment]
> > It seems I'm missing some privileges to resolve the conversations I've opened (seems weird). So I have LGTMed them.
> 
> 
> 
> @bobeal so from your perspective the PR is ready to be merged?

Yes!

File: doc/advanced-topics.md [hunk 4]
```diff
@@ -127,6 +127,58 @@ Other unrecognised `type` attributes will be passed as NGSI-LD data using the fo
     }
```
Comment (fgalan) [inline doc/advanced-topics.md:127]: Maybe the support to linked data is a relevant item to mention in the CHANGES_NEXT_RELEASE sublist at https://github.com/telefonicaid/iotagent-node-lib/blob/master/CHANGES_NEXT_RELEASE#L7-L14 ?

Comment (jason-fox) [reply doc/advanced-topics.md:185]: Fixed 891e514

**Review by fgalan — APPROVED** [review]
LGTM

**Comment by jason-fox** [comment]
@djs0109  - Yes, the `addLinkedEntities()` function is [here](https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/services/ngsi/entities-NGSI-LD.js#L363) which is a code path only run in `LD` mode - the same function could be ported or called in `v2` if a similar function was added to ``entities-NGSI-v2.js` - the path of execution for v2 and LD is separate.

**Comment by mrutid** [comment]
In not sure if i'm understanding correctly your needs, but In NGSIv2 I think it is a basic use case for **_multientity plugin_** where one single measure populates several entities. 

**In the following draft example measure with id 'p' will be provisioned at two different entities attributes:**

`Entitity[MyWeatherStation].Type[WeatherStation].Attribute[pressure].Type[Hgmm]
`AND
`Entitity[LinkedPressuerEntity].Type[linkedWeatherStation].Attribute[linked_pressure_attr].Type[Hgmm]
`
Given this Service/Device provisión (just a draft example, it may contain some mistakes):
```
{
            entity_name: 'MyWeatherStation',
            type: 'WeatherStation',
            active: [
                {
                    object_id: 'p',
                    name: 'pressure',
                    type: 'Hgmm'
                },
                {
                    object_id: 'foo_p',
                    name: 'linked_pressure_attr',
                    type: 'Hgmm',
                    expression:'p',
                    entity_name: 'LinkedPressuerEntity',
                    entity_type: 'linkedWeatherStation"
                }
            ]
        }
```

**Comment by djs0109** [comment]
Yes, the **_multientity plugin_** definitely helps. I am not aware before of this plugin. Thank you @mrutid

**Comment by mrutid** [comment]
Along with the expression (JEXL) plugin its a really powerful tool. I agree that docs doesn't help to discover it...

The "entity_name" in the attrs section may be a JEXL expression also.

**Pull request merged**